### PR TITLE
allow absolutely positioned popups to actually be anchored at the button

### PR DESCRIPTION
**Pull request by tobibeer**
e.g. those reveals in some $:/tags/PageTemplate, for example

Requires to specify at the button that the popup offset is to be calculated **absolute**, rather than relative (assuming the reveal to be but the next element right after the button).
## Demo

http://2128.tiddlyspot.com/

### CODE REVIEW

**Comment by Jermolene** [comment]
@tobibeer at a quick glance this looks reasonable. Can you add documentation and examples? It would be pretty cool if you could review and update the existing popup docs if you get a chance.

**Comment by tobibeer** [comment]
Would you be fine with button attribute as it is? Not sure if there may be another "position"ing mode, after all. In other words: saying `<$button popup="foo" absolute/>` works, but perhaps the more extensible choice is `<$button popup="foo" position:"absolute"/>`.

As for the docs, I can do that, just wanted to make sure I get your thumbs up first.

There is one thing I haven't quite understood yet.

If I create two tiddlers based on this PR:

```
title: Button

<$button popup="$:/SamplePopupState">Pop me up!</$button>
```

```
title: Reveal
tags: $:/tags/PageTemplate

<$reveal type="popup" state="$:/SamplePopupState">
<div class="tc-drop-down">

! This is the popup
And this is some text

</div>
</$reveal>
```

Now if I add either:

```
list-before: $:/core/ui/PageTemplate/story
```

or

```
list-after: $:/core/ui/PageTemplate/story
```

...to the **Reveal** then the positioning for the latter is as expected, but the former shows a small vertical distance downwards. I have not yet figured out where that's coming from.

**Comment by tobibeer** [comment]
It actually feels like I would want to define the positioning at the reveal, not the button, but then it's not as easy to have that attribute carried over to the button click. Could be saved to some field at the state, perhaps.

**Comment by tobibeer** [comment]
Added a demo link: http://2128.tiddlyspot.com

**Comment by pmario** [comment]
issue: This popup doesn't survive window resizing. 

To reproduce: 

- open the popup
- resize the window

**Comment by pmario** [comment]
It seems the problems are CSS media queries. Standard resizing seems to be ok. more tests needed.

**Comment by tobibeer** [comment]
@pmario, actually, no popup does survive resizing... in other words, a resize event should always update all popups with respect to their anchors.

**Comment by pmario** [comment]
If you type a search string in the TW search, the popup happily survives the resizing.

**Comment by pmario** [comment]
second problem:

- open the popup with your "Pop me up!" button
- select the search input area - your popup is still open
- type in a search word -> no popup anymore.

**Comment by tobibeer** [comment]
@pmario, I don't know how precisely the search "popup" on prerelease is "special". For comparison, try the last example in the RevealWidget.

>* select the tw search input area - your popup is still open
>* type in a search word -> no popup anymore.

this very much appears to be an issue with how this input is being handled, i.e. in how tries to toggle a the results popup, perhaps even when there is no result just yet.

I have added an example input created from the edit-text widget to the demo that shows you how clicking into it very much closes the popup.

So, it seems you have uncovered two more issues with popup handling, that appear unrelated to this PR:

1) no popup position refreshing upon resizing
2) no event bubbling for the search input all the way to whichever element handles closing popups
    * this element should possibly also be the one that will be given the resize event to update popup positions

@Jermolene was mentioning some inconsisitencies of how the search popup sometimes arbitrarily doesn't work as expected, perhaps all this brings us closer to understanding what that is

**Comment by pmario** [comment]
This PR has a problem! I'll record a video and upload it. link follows.

**Comment by tobibeer** [comment]
Hi Mario, thanks for taking the time to show the effect in a video. Then again, I clearly understood both issues when you first mentioned them and did reproduce both. However, I explained that (and how) both are very much unrelated to this PR ...at the bottom of [this comment](https://github.com/Jermolene/TiddlyWiki5/pull/2128#issuecomment-164778065).

**Comment by tobibeer** [comment]
@pmario , thanks for taking the time to show the effect(s) in a video. Then again, I clearly understood those issues when you first mentioned them and did reproduce both. However, I explained that (and how) both are very much **un**related to this PR ...in my [last comment](https://github.com/Jermolene/TiddlyWiki5/pull/2128#issuecomment-164778065).

@pmario / @Jermolene it would be nice of you to confirm the above, so we can...

* perhaps create one or two new issues for those
* concentrate on how to move on with this, e.g. if we're fine with:
    * how the attribute is named (**absolute** vs. **position:absolute**)
    * where to actually declare it (**button** vs. **reveal**)

**Comment by tobibeer** [comment]
@Jermolene, added the documentation.

I hope you find this can make it into 5.1.10. My [[AppearWidget|http://appear-demo.tiddlyspot.com/#Welcome]] would thus not need any hacky core amendments and, in fact, slightly buggy click handler overloading.

Both of @pmario's concerns are valid but actually different issues( ...that need creating). I'd appreciate if you get around taking another look at the search results popup issue (also mentioned above) before release. Would be a usability plus and help user acceptance of core changes regarding search.

**Comment by Jermolene** [comment]
Hi @tobibeer I think this is too controversial for 5.1.10 (see #2143).

Are there any lower level hooks that we could add for 5.1.10 that would help you avoid some hackiness?

**Comment by tobibeer** [comment]
@Jermolene,

The buggy hackiness takes place in these lines:

https://www.dropbox.com/sh/l8cembao5tavpjd/AADnADneZ9BgVD0DprA3fGrxa?dl=0

Perhaps there is a way to remove the original click handler?

The core ButtonWidget registers an anonymous function and as far as I could figure out, it cannot be unregistered so long as it is anonymous:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/button.js#L68

I guess if it was actually a function of the button widget, I would could perhaps address it / remove it and replace it with mine.

**Comment by tobibeer** [comment]
@Jermolene,

The buggy hackiness takes place in these lines:

https://github.com/tobibeer/tw5-appear/blob/master/src/appear.js#L147-L150

Perhaps there is a way to remove the original click handler?

The core ButtonWidget registers an anonymous function and as far as I could figure out, it cannot be unregistered so long as it is anonymous:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/button.js#L68

I guess if it was actually a function of the button widget, I would could perhaps address it / remove it and replace it with mine.

**Comment by tobibeer** [comment]
> Hi @tobibeer I think this is too controversial for 5.1.10 (see #2143).

May I ask what you find controversial about this PR?

**Comment by Jermolene** [comment]
Hi @tobibeer 

> May I ask what you find controversial about this PR?

It may be fine, but I can't tell without investing more time whether it's a good, general solution, and whether the implementation is solid. I'd like to spend more time to look at your examples.

There's also the resizing problem that @pmario identified. The current core approach of expecting the button and the reveal widget to share a coordinate system does permit popups to be made that automatically stay correctly positioned during resizing.

I'm sorry it's not made 5.1.10. The point is that I've only got a few hours left before we release, so I can't afford to merge pull requests that need more than a cursory verification.

**Comment by tobibeer** [comment]
> The current core approach of expecting the button and the reveal widget to share a coordinate system does permit popups to be made that automatically stay correctly positioned during resizing.

That does not appear to be true, as I have tried to point out... because there is **no** recalculation. The effect (of popups being mal-positioned) may not be as dramatic with "in-flow" popups, but it's still there.

I'm fine with it not making it into 5.1.10 and I think that fixing the set widget alone (for me) is significant enough for 5.1.10... although there's a lot more to it than that. :D

**Comment by Jermolene** [comment]
> because there is no recalculation

The whole problem of repositioning popups is largely avoided in the core by judicious use of shared coordinate systems. It's not clear to me without further investigation which cases are failing, and whether your proposal is the best solution.

**Comment by tobibeer** [comment]
@Jermolene, I tried to factor the click handler out of the **render** function but didn't quite get there with respect to keeping a reference to "self" alive.

> The core ButtonWidget registers an anonymous function and as far as I could figure out, it cannot be unregistered so long as it is anonymous:
> https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/button.js#L68

Any chance you'd know how to create some working version of a ...

`ButtonWidget.prototype.click = function(event) {}`

...and have that be assigned to the domNode as the click handler, so that it can later be removed via `domNode.removeEventHandler()`?

**Comment by tobibeer** [comment]
@Jermolene, I have decided to remove the hackiness by "properly" [hijacking Popup.show](https://github.com/tobibeer/tw5-appear/blob/master/src/appear.js#L393-L440).

File: core/modules/utils/dom/popup.js [hunk 1]
```diff
@@ -109,28 +109,45 @@ Display a popup by adding it to the stack
...
 	if(this.popups.length > 0) {
-		this.rootElement.addEventListener("click",this,true);		
+		this.rootElement.addEventListener("click",this,true);
 	}
 };
 
 /*
+Calculates absolute offset of anchor in document
+*/
+Popup.prototype.getAbsoluteOffset = function(element) {
```
Comment (Jermolene) [inline core/modules/utils/dom/popup.js]: This function should be in utils/dom.js

File: editions/tw5.com/tiddlers/widgets/ButtonWidget.tid [hunk 2]
```diff
@@ -29,6 +29,7 @@ The content of the `<$button>` widget is displayed within the button.
 |set |A TextReference to which a new value will be assigned |
 |setTo |The new value to assign to the TextReference identified in the `set` attribute |
 |popup |Title of a state tiddler for a popup that is toggled when the button is clicked |
+|absolute |When specified, calculates the popup offset absolute, not relative, i.e. whenever the [[reveal|RevealWidget]] is not the next element of the button|
```
Comment (Jermolene) [inline editions/tw5.com/tiddlers/widgets/ButtonWidget.tid]: Documentation updated needs to be tagged with the version

**Comment by pmario** [comment]
As mentioned in the hangout, I did find some problems with the core reveal widget, that go down to the .js implementation. ... I'll need to find the tiddlywiki, with which I did some tests. ... So we may fix the core first.

**Comment by tobibeer** [comment]
@Jermolene, resolved conflicts and moved getAbsoluteOffset to dom utils and added version to docs

**Comment by pmario** [comment]
@Jermolene ... Popup handling has completely changed, since this PR was created. I think it's safe to close it

**Comment by Jermolene** [comment]
Thanks @pmario

**Pull request closed without merging**